### ISSUE TITLE
math.simplify loses custom functions

### ISSUE THREAD

**Issue by IvanVergiliev** [josdejong/mathjs#916]
I'm trying to create a custom sigmoid function using a `FunctionAssignmentNode`, but after I run simplify on an expression containing it, the function definition is lost and so `eval` later fails with `Undefined symbol sigma`.

Sample code:
```js
> const s = new math.expression.node.FunctionAssignmentNode('sigma', ['x'], math.parse('1 / (1 + exp(-x))'));
> const f = new node.FunctionNode(s, [new node.SymbolNode('x')]);
> f.toString()
'sigma(x) = 1 / (1 + exp(-x))(x)'
> f.eval({x: 5})
0.9933071490757153
```
So far, so good. However, after I run `simplify`, things stop working:
```js
> const fSimplified = math.simplify(f)
> fSimplified.toString()
'sigma(x)'
> fSimplified.eval({x: 5})
Error: Undefined symbol sigma
    at undef (/Users/ivan/src/backprop/node_modules/mathjs/lib/expression/node/SymbolNode.js:120:11)
    [...]
```

**Comment by firepick1** [comment]
I reproduced your example in a unit test and will be looking at it. Thanks for the very clear example!

**Comment by firepick1** [comment]
I checked a fix into simplify2 branch. I've never worked with FunctionAssignmentNodes, so the fix is very conservative and no simplification is done. From this baseline we can get better as needed. @jos review please.

**Comment by josdejong** [comment]
Thanks for your fix @firepick1, this indeed solves the immediate issue, though there is still an underlying issue. Reason is that a FunctionNode mostly refers to a SymbolNode (which has a name), but that's not always the case:
```js
sin(2)                  // a FunctionNode pointing to a SymbolNode `sin`

someObject.toString()   // a FunctionNode pointing to an AccessorNode (accessing 
                        // the method toString on someObject)

(f(x) = x^2)(4)         // a FunctionNode pointing to a FunctionAsssignmentNode 
                        // (immediately invoked)
```

So I think the reason is that `node.name` that was used in the simplify part for FunctionNode, but this property *only* exists when node.fn is a SymbolNode. Does that make sense?

I've solved it in https://github.com/josdejong/mathjs/commit/54230e8dc92b29dd5b002ac58ef850ac1417c9cb I think, but please have a look at it.

**Comment by firepick1** [comment]
Thanks for your deeper understanding and simple edit. Looks fine--that was probably leftover from a random meandering of mine.

**Comment by josdejong** [comment]
:+1:

**Comment by josdejong** [comment]
Should be fixed now in `v3.16.1`

**Comment by IvanVergiliev** [comment]
Thanks for the fix!
